File: structures/stack_linked.c
// implementation of integer stack using linked list 

#include <stdio.h>
#include <stdlib.h>

typedef struct stackNode {
    int  data;
    struct stackNode *next;
}
stackNode;

struct stackNode *newNode(int data);
void push(stackNode **stack, int data);
int pop(stackNode **stack);
int peek(stackNode *stack);

int main(void) {
    stackNode *stack = NULL;
    push(&stack, 2);
    printf("%d\n", stack->data);
    push(&stack, 5);
    printf("%d\n", stack->data);
    printf("%d\n", pop(&stack));
    printf("%d\n", peek(stack));
    printf("%d\n", pop(&stack));
    printf("%d\n", pop(&stack));
    printf("%d\n", pop(&stack));
    printf("%d\n", peek(stack));
    free(stack);
}

struct stackNode *newNode(int data) {
    stackNode *stack = malloc(sizeof(stackNode));
    if (stack != NULL) {
        stack->data = data;
        stack->next = NULL;
    }
    return stack;    
}

void push(stackNode **stack, int data) {
    stackNode *new = newNode(data); 
    new->next = *stack;
    *stack = new;
}

int pop(stackNode **stack) {
    if (*stack == NULL) {
        return -1234567;
    }
    stackNode *tmp = *stack;
    *stack = tmp->next;
    int popped = tmp->data;
    free(tmp);
    return(popped);     
}

int peek(stackNode *stack) {
    if (stack == NULL) {
        return -1234567;
    }
    return stack->data;
}

File: structures/stack_linked.py
class Node:
    def __init__(self, data = None, next = None):
        self.data = data
        self.next = next

    def __str__(self):
        return str(self.data)

    def getData(self):
        return self.data
    
    def setData(self, newData):
        self.data = newData
    
    def getNext(self):
        return self.next
    
    def setNext(self, newNext):
        self.next = newNext

class Stack:
    def __init__(self, head = None, tail = None):
        self.head = head
        self.tail = tail

    def push(self, node):
        node.setNext(self.tail)
        self.tail = node
        return

    def pop(self):
        if self.tail == None:
            return None
        tmp = self.tail
        self.tail = tmp.getNext() 
        return tmp.getData()

    def peek(self): 
        return self.tail.getData()

if __name__ == "__main__":
    myStack = Stack()
    myStack.push(Node(1))
    print(myStack.peek())
    myStack.push(Node(2))
    print(myStack.peek())
    myStack.push(Node(3))
    print(myStack.peek())
    print(myStack.pop())
    print(myStack.pop())
    print(myStack.pop())
    print(myStack.pop())